File: android/src/com/android/tools/idea/gradle/structure/daemon/PsLibraryUpdateCheckerDaemon.java
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.gradle.structure.daemon;

import com.android.ide.common.repository.GradleVersion;
import com.android.tools.idea.gradle.structure.configurables.PsContext;
import com.android.tools.idea.gradle.structure.model.PsArtifactDependencySpec;
import com.android.tools.idea.gradle.structure.model.PsLibraryDependency;
import com.android.tools.idea.gradle.structure.model.android.PsAndroidModule;
import com.android.tools.idea.gradle.structure.model.repositories.search.ArtifactRepository;
import com.android.tools.idea.gradle.structure.model.repositories.search.FoundArtifact;
import com.android.tools.idea.gradle.structure.model.repositories.search.SearchRequest;
import com.android.tools.idea.gradle.structure.model.repositories.search.SearchResult;
import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import com.google.common.util.concurrent.Futures;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.application.Application;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.util.EventDispatcher;
import com.intellij.util.ui.update.MergingUpdateQueue;
import com.intellij.util.ui.update.Update;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.concurrent.Future;
import java.util.concurrent.atomic.AtomicBoolean;

import static com.intellij.openapi.util.text.StringUtil.isNotEmpty;
import static com.intellij.util.ui.update.MergingUpdateQueue.ANY_COMPONENT;

public class PsLibraryUpdateCheckerDaemon extends PsDaemon {
  @NotNull private final MergingUpdateQueue myMainQueue;
  @NotNull private final MergingUpdateQueue myResultsUpdaterQueue;

  @NotNull private final AvailableLibraryUpdates myResults = new AvailableLibraryUpdates();
  @NotNull private final AtomicBoolean myRunning = new AtomicBoolean(true);

  @NotNull private final EventDispatcher<AvailableUpdatesListener> myEventDispatcher =
    EventDispatcher.create(AvailableUpdatesListener.class);

  public PsLibraryUpdateCheckerDaemon(@NotNull PsContext context) {
    super(context);
    myMainQueue = createQueue("Project Structure Daemon Update Checker", null);
    myResultsUpdaterQueue = createQueue("Project Structure Available Update Results Updater", ANY_COMPONENT);
  }

  public void queueUpdateCheck() {
    myMainQueue.queue(new SearchForAvailableUpdates());
  }

  @Override
  @NotNull
  protected MergingUpdateQueue getMainQueue() {
    return myMainQueue;
  }

  @Override
  @NotNull
  protected MergingUpdateQueue getResultsUpdaterQueue() {
    return myResultsUpdaterQueue;
  }

  @NotNull
  public AvailableLibraryUpdates getResults() {
    return myResults;
  }

  public void add(@NotNull AvailableUpdatesListener listener, @NotNull Disposable parentDisposable) {
    myEventDispatcher.addListener(listener, parentDisposable);
  }

  public boolean isRunning() {
    return myRunning.get();
  }

  private void search(@NotNull Collection<ArtifactRepository> repositories,
                      @NotNull Collection<LibraryUpdateId> ids) {
    myRunning.set(true);
    myResults.clear();

    int resultCount = repositories.size() * ids.size();
    List<Future<SearchResult>> jobs = Lists.newArrayListWithExpectedSize(resultCount);

    Set<SearchRequest> requests = Sets.newHashSet();
    ids.forEach(id -> {
      SearchRequest request = new SearchRequest(id.getName(), id.getGroupId(), 1, 0);
      requests.add(request);
    });

    Set<SearchResult> results = Sets.newHashSet();
    List<Exception> errors = Lists.newArrayList();

    Application application = ApplicationManager.getApplication();
    application.executeOnPooledThread(() -> {
      for (ArtifactRepository repository : repositories) {
        for (SearchRequest request : requests) {
          jobs.add(application.executeOnPooledThread(() -> repository.search(request)));
        }
      }

      for (Future<SearchResult> job : jobs) {
        try {
          SearchResult result = Futures.get(job, Exception.class);
          List<FoundArtifact> artifacts = result.getArtifacts();
          if (artifacts.size() == 1) {
            FoundArtifact artifact = artifacts.get(0);
            if (!artifact.getVersions().isEmpty()) {
              results.add(result);
            }
          }
        }
        catch (Exception e) {
          errors.add(e);
        }
      }

      for (SearchResult result : results) {
        List<FoundArtifact> artifacts = result.getArtifacts();
        myResults.add(artifacts.get(0));
      }

      myResultsUpdaterQueue.queue(new UpdatesAvailable());
    });
  }

  private class SearchForAvailableUpdates extends Update {
    public SearchForAvailableUpdates() {
      super(getContext().getProject());
    }

    @Override
    public void run() {
      Set<ArtifactRepository> repositories = Sets.newHashSet();
      Set<LibraryUpdateId> ids = Sets.newHashSet();

      getContext().getProject().forEachModule(module -> {
        repositories.addAll(module.getArtifactRepositories());

        if (module instanceof PsAndroidModule) {
          PsAndroidModule androidModule = (PsAndroidModule)module;

          androidModule.forEachDeclaredDependency(dependency -> {
            if (dependency instanceof PsLibraryDependency) {
              PsLibraryDependency libraryDependency = (PsLibraryDependency)dependency;
              PsArtifactDependencySpec spec = libraryDependency.getDeclaredSpec();
              if (spec != null && isNotEmpty(spec.version)) {
                GradleVersion version = GradleVersion.tryParse(spec.version);
                if (version != null) {
                  ids.add(new LibraryUpdateId(spec));
                }
              }
            }
          });
        }
      });

      if (repositories.size() > 0 && ids.size() > 0) {
        search(repositories, ids);
      }
    }
  }

  private class UpdatesAvailable extends Update {
    public UpdatesAvailable() {
      super(getContext().getProject());
    }

    @Override
    public void run() {
      myRunning.set(false);
      myEventDispatcher.getMulticaster().availableUpdates();
    }
  }

  public interface AvailableUpdatesListener extends EventListener {
    void availableUpdates();
  }
}

File: android/src/com/android/tools/idea/gradle/structure/daemon/PsLibraryUpdateCheckerDaemon.kt
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.gradle.structure.daemon

import com.android.annotations.concurrency.UiThread
import com.android.ide.common.repository.GradleVersion
import com.android.tools.idea.gradle.structure.configurables.PsContext
import com.android.tools.idea.gradle.structure.configurables.RepositorySearchFactory
import com.android.tools.idea.gradle.structure.daemon.AvailableLibraryUpdateStorage.AvailableLibraryUpdates
import com.android.tools.idea.gradle.structure.model.PsProject
import com.android.tools.idea.gradle.structure.model.android.PsAndroidModule
import com.android.tools.idea.gradle.structure.model.repositories.search.ArtifactRepository
import com.android.tools.idea.gradle.structure.model.repositories.search.SearchQuery
import com.android.tools.idea.gradle.structure.model.repositories.search.SearchRequest
import com.android.tools.idea.gradle.structure.model.repositories.search.getResultSafely
import com.intellij.openapi.Disposable
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.util.text.StringUtil.isNotEmpty
import com.intellij.util.EventDispatcher
import com.intellij.util.ui.UIUtil
import com.intellij.util.ui.update.MergingUpdateQueue
import com.intellij.util.ui.update.MergingUpdateQueue.ANY_COMPONENT
import com.intellij.util.ui.update.Update
import java.util.EventListener
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.function.Consumer

private val LOG = Logger.getInstance(PsLibraryUpdateCheckerDaemon::class.java)

class PsLibraryUpdateCheckerDaemon(
  parentDisposable: Disposable,
  private val project: PsProject,
  private val repositorySearchFactory: RepositorySearchFactory
) : PsDaemon(parentDisposable) {
  override val mainQueue: MergingUpdateQueue = createQueue("Project Structure Daemon Update Checker", null)
  override val resultsUpdaterQueue: MergingUpdateQueue = createQueue("Project Structure Available Update Results Updater", ANY_COMPONENT)

  private val eventDispatcher = EventDispatcher.create(AvailableUpdatesListener::class.java)

  fun getAvailableUpdates(): AvailableLibraryUpdates = AvailableLibraryUpdateStorage.getInstance(project.ideProject).getState()

  fun queueAutomaticUpdateCheck() {
    val searchTimeMillis = getAvailableUpdates().lastSearchTimeMillis
    if (searchTimeMillis > 0) {
      val elapsed = System.currentTimeMillis() - searchTimeMillis
      val daysPastSinceLastUpdate = TimeUnit.MILLISECONDS.toDays(elapsed)
      if (daysPastSinceLastUpdate < 3) {
        // Display stored updates from previous search
        resultsUpdaterQueue.queue(UpdatesAvailable())
        return
      }
    }
    queueUpdateCheck()
  }

  private fun queueUpdateCheck() {
    mainQueue.queue(SearchForAvailableUpdates())
  }

  fun add(@UiThread listener: () -> Unit, parentDisposable: Disposable) {
    eventDispatcher.addListener(
      object : AvailableUpdatesListener {
        override fun availableUpdates() = listener()
      }, parentDisposable)
  }

  private fun search(
    repositories: Collection<ArtifactRepository>,
    ids: Collection<LibraryUpdateId>
  ) {
    getAvailableUpdates().clear()

    val requests =
      ids
        .map { id -> SearchRequest(SearchQuery(id.groupId, id.name), 1, 0) }
        .toSet()

    val searcher = repositorySearchFactory.create(repositories)
    val resultFutures = requests.map { searcher.search(it) }

    Disposer.register(this, Disposable {
      resultFutures.forEach { it.cancel(true) }
    })

    val foundArtifacts =
      resultFutures
        .map { it.getResultSafely() }
        .flatMap { it?.artifacts.orEmpty() }

    val updates = getAvailableUpdates()
    foundArtifacts.forEach { updates.add(it) }
    updates.lastSearchTimeMillis = System.currentTimeMillis()

    resultsUpdaterQueue.queue(UpdatesAvailable())
  }

  private inner class SearchForAvailableUpdates : Update(project) {
    override fun run() {
      val repositories = mutableSetOf<ArtifactRepository>()
      val ids = mutableSetOf<LibraryUpdateId>()
      UIUtil.invokeAndWaitIfNeeded(Runnable {
        project.forEachModule(Consumer { module ->
          repositories.addAll(module.getArtifactRepositories())
          if (module is PsAndroidModule) {
            module.dependencies.forEachLibraryDependency { dependency ->
              val spec = dependency.spec
              if (isNotEmpty(spec.version)) {
                val version = GradleVersion.tryParse(spec.version!!)
                if (version != null) {
                  ids.add(LibraryUpdateId(spec.group.orEmpty(), spec.name))
                }
              }
            }
          }
        })
      })
      if (repositories.isNotEmpty() && ids.isNotEmpty()) {
        search(repositories, ids)
      } else {
        resultsUpdaterQueue.queue(UpdatesAvailable())
      }
    }
  }

  private inner class UpdatesAvailable : Update(project) {

    @UiThread
    override fun run() {
      eventDispatcher.multicaster.availableUpdates()
    }
  }

  private interface AvailableUpdatesListener : EventListener {
    @UiThread
    fun availableUpdates()
  }
}
